src/server/service/kv.rs

```diff
@@ -2265,6 +2275,32 @@ fn future_raw_coprocessor<E: Engine, L: LockManager, F: KvFormat>(
     async move { Ok(ret.await) }
 }
 
+fn future_broadcast_txn_status<E: Engine, L: LockManager, F: KvFormat>(
+    storage: &Storage<E, L, F>,
+    req: BroadcastTxnStatusRequest,
+) -> impl Future<Output = ServerResult<BroadcastTxnStatusResponse>> {
+    let cache = storage.get_scheduler().get_txn_status_cache();
+    async move {
+        let now = SystemTime::now();
+        for txn_status in req.txn_status.iter() {
+            let txn_state = TxnState::from_ts(
+                txn_status.start_ts.into(),
+                txn_status.min_commit_ts.into(),
+                txn_status.commit_ts.into(),
+                txn_status.rolled_back,
+            );
+            if txn_status.is_completed {
+                // large_txn_cache is only for **ongoing** large txns, so remove it when
+                // completed.
+                assert_matches!(txn_state, TxnState::Committed { .. } | TxnState::RolledBack);
+                cache.remove_large_txn(txn_status.start_ts.into());
+            }
+            cache.upsert(txn_status.start_ts.into(), txn_state, now);
```

Comment:
The cache is possible blocking operation, this task should be handled by the scheduler worker pool and avoid its impact on the gRPC threads.

Reply:
Good catch